crates/calibration/src/residuals.rs

```diff
@@ -0,0 +1,107 @@
+            Some(Vec::new()),
+            |residuals: Option<Vec<f32>>, measurements| match residuals {
+                Some(mut residuals) => {
+                    residuals.extend::<Vec<f32>>(
+                        Residuals::calculate_from(parameters, measurements, field_dimensions)
+                            .ok()?
+                            .into(),
+                    );
+                    Some(residuals)
+                }
+                None => None,
+            },
+        )
+        .map(DVector::from_vec)
+}
+
+pub struct Residuals {
+    pub border_to_connecting_angle: f32,
+    pub connecting_to_goal_box_angle: f32,
+    pub distance_between_parallel_line_start_points: f32,
+    pub distance_between_parallel_line_center_points: f32,
+    pub distance_between_parallel_line_end_points: f32,
+}
+
+impl Residuals {
+    pub fn calculate_from(
+        parameters: &Corrections,
+        measurement: &Measurement,
+        field_dimensions: &FieldDimensions,
+    ) -> Result<Self, ResidualsError> {
+        let corrected = measurement.matrix.to_corrected(
+            parameters.correction_in_robot,
+            match measurement.position {
+                CameraPosition::Top => parameters.correction_in_camera_top,
+                CameraPosition::Bottom => parameters.correction_in_camera_bottom,
+            },
+        );
+
+        let projected_lines = measurement
+            .lines
+            .to_projected(&corrected)
+            .map_err(ResidualsError::NotProjected)?;
+
+        let border_to_connecting_angle = projected_lines
+            .border_line
+            .angle(projected_lines.connecting_line);
+        let connecting_to_goal_box_angle = projected_lines
+            .border_line
+            .angle(projected_lines.connecting_line);
+        let distance_between_parallel_line_start_points = projected_lines
+            .border_line
+            .distance_to_point(projected_lines.goal_box_line.0);
+        let distance_between_parallel_line_center_points = projected_lines
+            .border_line
+            .distance_to_point(projected_lines.goal_box_line.center());
+        let distance_between_parallel_line_end_points = projected_lines
+            .border_line
+            .distance_to_point(projected_lines.goal_box_line.1);
+
+        Ok(Residuals {
+            border_to_connecting_angle: border_to_connecting_angle - FRAC_PI_2,
+            connecting_to_goal_box_angle: connecting_to_goal_box_angle - FRAC_PI_2,
+            distance_between_parallel_line_start_points: distance_between_parallel_line_start_points
+                - field_dimensions.goal_box_area_length,
+            distance_between_parallel_line_center_points:
+                distance_between_parallel_line_center_points - field_dimensions.goal_box_area_length,
+            distance_between_parallel_line_end_points: distance_between_parallel_line_end_points
+                - field_dimensions.goal_box_area_length,
```

Comment:
The residuals here are measured in different units and ranges. You might want to normalize them or add different weights for each one of them.

Reply:
What I recall from some papers about the solver (Levenberg-Marquardt) is that it is relatively immune to this issue, the ranges of each element of residual vector does not have to be  the same. As I remember the reason is the way it calculates the relative reduction of error - not as a norm or sum but it is done element-wise.

However, as you suggested, weights or some other influencing factors can be introduced to Influence which parameter gets priority, then normalizing would probably help to tune that.

Reply:
Ok, I will have a look once I conduct more in-depth testing with this calibration problem. I would like to keep it as is for now.